httpclient/src/main.rs

```diff
@@ -0,0 +1,179 @@
+                &req.method, &req.url, status_code
+            );
+        }
+        Ok(HttpResponse {
+            body: body.to_vec(),
+            header: headers,
+            status_code,
+        })
+    }
+}
+
+/// convert response headers from reqwest to HeaderMap
+fn convert_response_headers(
+    headers: &reqwest::header::HeaderMap,
+) -> wasmcloud_interface_httpclient::HeaderMap {
+    let mut hmap = wasmcloud_interface_httpclient::HeaderMap::new();
+    for k in headers.keys() {
+        let vals = headers
+            .get_all(k)
+            .iter()
+            // from http crate:
+            //    In practice, HTTP header field values are usually valid ASCII.
+            //     However, the HTTP spec allows for a header value to contain
+            //     opaque bytes as well.
+            // This implementation only forwards headers with ascii values to the actor.
+            .filter_map(|val| val.to_str().ok())
+            .map(|s| s.to_string())
+            .collect::<Vec<_>>();
+        if !vals.is_empty() {
+            hmap.insert(k.to_string(), vals);
+        }
+    }
+    hmap
+}
+
+/// convert HeaderMap from actor into outgoing HeaderMap
+fn convert_request_headers(
+    header: &wasmcloud_interface_httpclient::HeaderMap,
+    headers_mut: &mut http::HeaderMap,
+) {
+    let map = headers_mut;
+    for (k, vals) in header.iter() {
+        let name = match http::HeaderName::from_bytes(k.as_bytes()) {
+            Ok(name) => name,
+            Err(e) => {
+                error!(
+                    "invalid response header name: '{}': {} - sending without this header",
+                    &k,
+                    &e.to_string()
+                );
+                continue;
+            }
+        };
+        for val in vals.iter() {
+            let value = match http::HeaderValue::from_str(val) {
+                Ok(value) => value,
+                Err(e) => {
+                    error!(
+                        "Non-ascii header value: '{}': {} - skipping this header",
+                        &val,
+                        &e.to_string()
+                    );
+                    continue;
+                }
+            };
+            map.append(&name, value);
+        }
+    }
+}
+
+/// Handle incoming rpc messages and dispatch to applicable trait handler.
+#[async_trait]
+impl MessageDispatch for HttpClientProvider {
+    async fn dispatch(&self, ctx: &Context, message: Message<'_>) -> RpcResult<Message<'_>> {
+        let op = match message.method.split_once('.') {
+            Some((cls, op)) if cls == "HttpClient" => op,
+            None => message.method,
+            _ => {
+                return Err(RpcError::MethodNotHandled(message.method.to_string()));
+            }
+        };
+        HttpClientReceiver::dispatch(
+            self,
+            ctx,
+            &Message {
+                method: op,
+                arg: message.arg,
+            },
+        )
+        .await
```

Comment:
Is this part codegenned?

Reply:
not yet :-). There are still a couple of open issues that need to be nailed down before this can be code-genned:
- can a capability provider implement more than one interface? (I think yes)
- should messages have a "Service." prefix?
The latter is something I was going to test today. I believe actors and the providers I've built have support for it. The only thing that doesn't is the builtin providers.

Reply:
Okay, I was just wondering here 🙂 implementing more than one service is definitely something to discuss